arnglld/src/main.rs

```diff
@@ -33,19 +40,175 @@ struct Opt {
+                if dev_name == lc_name {
+                    return Ok(device);
+                }
+            }
+
+            // Then try prefixes
+            for device in host.input_devices()? {
+                let dev_name = device.name()?.to_lowercase();
+                if dev_name.starts_with(&lc_name) {
+                    return Ok(device);
+                }
+            }
+
+            // Then try for any device that matches the substring
+            for device in host.input_devices()? {
+                let dev_name = device.name()?.to_lowercase();
+                if dev_name.contains(&lc_name) {
+                    return Ok(device);
+                }
+            }
+
+            return Err(format_err!("Cannot find input device matching {:?}", name));
+        }
+
+        host.default_input_device()
+            .ok_or_else(|| format_err!("no default input device"))
+    }
+
+    #[allow(dead_code)]
+    fn get_packet_stream(&self) -> Result<Bell202Receiver, anyhow::Error> {
+        let device = self.get_input_device()?;
+        info!("Using input device {:?}", device.name());
+        let receiver = Bell202Receiver::new(&device)?;
+
+        Ok(receiver)
+    }
+
+    fn get_packet_sink(&self) -> Result<Bell202Sender, anyhow::Error> {
+        let device = self.get_output_device()?;
+        info!("Using output device {:?}", device.name());
+        let sender = Bell202Sender::new(&device)?;
+
+        Ok(sender)
+    }
 }
 
 fn main() {
     let opt = Opt::parse();
 
-    println!("Callsign: {}", opt.callsign);
-    println!("opt = {:?}", opt);
+    if opt.list_devices {
+        let host = cpal::default_host();
+        println!("Input Devices:");
+        for device in host.input_devices().expect("Unable to list input devices") {
+            if let Some(name) = device.name().ok() {
+                println!("\t{}", name);
+            }
+        }
+        println!("Output Devices:");
+        for device in host
+            .output_devices()
+            .expect("Unable to list output devices")
+        {
+            if let Some(name) = device.name().ok() {
+                println!("\t{}", name);
+            }
+        }
+        return;
+    }
 
     stderrlog::new()
         .quiet(opt.quiet)
         .verbosity(opt.verbose)
         .init()
         .unwrap();
+
+    println!("Callsign: {}", opt.callsign.expect("Missing callsign"));
+    println!("opt = {:?}", opt);
+
+    const X25: crc::Crc<u16> = crc::Crc::<u16>::new(&crc::CRC_16_IBM_SDLC);
+
+    let frame = "\x01\x00de N6DRC: TEST: This is a test frame of ASCII text."
+        .bytes()
+        .append_crc(&X25);
+
+    let mut packet_sink = opt.get_packet_sink().unwrap();
+
+    println!("Sending test frame...");
+
+    // Play the test packet.
+    block_on(packet_sink.send(frame.collect())).unwrap();
+
+    // Sleep while we play the packet in the background.
+    std::thread::sleep(std::time::Duration::from_secs(1));
```

Comment:
It'd be nice to be able to join on something instead of sleeping.

Reply:
Well, this delay here wasn't really doing anything useful anyway. Removing it should be fine.